nexus/src/app/instance.rs

```diff
@@ -1995,194 +1854,55 @@ impl super::Nexus {
-            info!(log, "vmm is terminated, cleaning up resources";
-                    "instance_id" => %instance_id,
-                    "propolis_id" => %propolis_id);
-
-            datastore
-                .sled_reservation_delete(opctx, propolis_id.into_untyped_uuid())
-                .await?;
-
-            if !datastore.vmm_mark_deleted(opctx, &propolis_id).await? {
-                warn!(log, "failed to mark vmm record as deleted";
-                    "instance_id" => %instance_id,
-                    "propolis_id" => %propolis_id,
-                    "vmm_state" => ?new_runtime_state.vmm_state);
-            }
-        }
-    }
-
-    match result {
-        Ok(result) => {
-            info!(log, "instance and vmm updated by sled agent";
-                    "instance_id" => %instance_id,
-                    "propolis_id" => %propolis_id,
-                    "instance_updated" => result.instance_updated,
-                    "vmm_updated" => result.vmm_updated,
-                    "migration_updated" => ?result.migration_updated);
-            Ok(Some(result))
-        }
-
-        // The update command should swallow object-not-found errors and
-        // return them back as failures to update, so this error case is
-        // unexpected. There's no work to do if this occurs, however.
-        Err(Error::ObjectNotFound { .. }) => {
-            error!(log, "instance/vmm update unexpectedly returned \
-                    an object not found error";
-                    "instance_id" => %instance_id,
-                    "propolis_id" => %propolis_id);
-            Ok(None)
-        }
+        .await?;
 
-        // If the datastore is unavailable, propagate that to the caller.
-        // TODO-robustness Really this should be any _transient_ error.  How
-        // can we distinguish?  Maybe datastore should emit something
-        // different from Error with an Into<Error>.
-        Err(error) => {
-            warn!(log, "failed to update instance from sled agent";
-                    "instance_id" => %instance_id,
-                    "propolis_id" => %propolis_id,
-                    "error" => ?error);
-            Err(error)
-        }
+    // If an instance-update saga must be executed as a result of this update,
+    // prepare and return it.
+    if instance_update::update_saga_needed(
+        &opctx.log,
+        instance_id,
+        new_runtime_state,
+        &result,
+    ) {
+        let (.., authz_instance) = LookupPath::new(&opctx, datastore)
+            .instance_id(instance_id.into_untyped_uuid())
+            .lookup_for(authz::Action::Modify)
+            .await?;
+        let saga = instance_update::SagaInstanceUpdate::prepare(
+            &instance_update::Params {
+                serialized_authn: authn::saga::Serialized::for_opctx(opctx),
+                authz_instance,
+            },
+        )?;
+        Ok(Some(saga))
+    } else {
+        Ok(None)
     }
```

Comment:
`notify_instance_updated` does a *lot* less work now: it just writes the VMM and migration updates to CRDB, decides if an instance update saga should run, and starts that saga.